### PR TITLE
ASoC: SOF: trace: fix trace doesn't update issue

### PR DESCRIPTION
**Pull request by keyonjie**
We have issues for DMA trace if runtime suspend is turned on:
a. Those logs before suspend will be cleared/overwritten after resume.
b. If we run logger at suspend, logs won't be updated though the ADSP is
running after that.

To fix these issues, this commit change the DMA trace behavior as below:

1. move init_waitqueue_head() to snd_sof_init_trace(), only
init_waitqueue() at first boot.
2. add a member host_init_offset to store the last trace offset before
suspend, and calculate the available log size based on it.
3. add a function hda_dsp_trace_setup_bdl() to setup bdl entries for DMA
trace, when host_init_offset!=0, we divide buffer into 2 BDL entries:
[host_init_offset, buffer_size) and [0, host_init_offset). This make FW
possible to copy new logs to destination from host_init_offset, which
won't overwrite the existed logs before suspend.

With these changes, both issues mentioned above are resolved on APL.

Todo: for SKL- platforms where not using BDL entries, FW to support
non-0 offset need to be implemented for those platforms, please refer
to this:
https://github.com/thesofproject/sof/issues/544

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/intel/hda.h [hunk 1]
```diff
@@ -343,6 +343,12 @@ struct sof_intel_dsp_desc {
 	struct snd_sof_dsp_ops *ops;
 };
 
+enum sof_hda_tream_type {
+	SOF_HDA_NORMAL_STREAM = 0,	/* normal playback/capture stream */
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda.h]: SOF_HDA_AUDIO_STREAM

File: sound/soc/sof/sof-priv.h [hunk 2]
```diff
@@ -363,6 +363,7 @@ struct snd_sof_dev {
 	struct snd_dma_buffer dmatp;
 	int dma_trace_pages;
 	wait_queue_head_t trace_sleep;
+	u32 host_init_offset; /* restore trace offset upon resume */
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h]: is host_init_offset and host_offset both related to trace only ? If can we change the word host-> trace

File: sound/soc/sof/intel/hda.h (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h]: fine, let me change.

File: sound/soc/sof/sof-priv.h (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h]: yes, they are both for trace only. let me change them.

File: sound/soc/sof/intel/hda-stream.c [hunk 3]
```diff
@@ -85,6 +85,60 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
+	} else {
+		/*
+		 * for first boot or trace_init_offset=0 cases, we need
+		 * assign only one BDL entry for DMA trace
+		 */
+		offset = hda_setup_bdle(sdev, dmab,
+					stream, &bdl, 0,
+					stream->bufsize, 0);
+		if (offset != stream->bufsize) {
+			dev_err(sdev->dev, "error: bdl setting for dma trace wrong!\n");
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-stream.c]: Btw, how is the value "wrong". Can you make the error messages more helpful for debug.

File: sound/soc/sof/trace.c [hunk 4]
```diff
@@ -154,16 +163,24 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 	if (sdev->dtrace_is_enabled)
 		return -EINVAL;
 
+	/* update init_offset, and reinitial offset to 0 */
+	sdev->trace_init_offset = (sdev->trace_init_offset +
+				  sdev->trace_offset) %
+				 sdev->dmatb.bytes;
```
Comment (ranj063) [inline sound/soc/sof/trace.c]: @keyonjie is this correct? should this just be 
sdev->trace_init_offset = sdev->trace_offset % sdev->dmatb.bytes?

trace_init_offset never gets reset so I am concerned this might skew the offset.

File: sound/soc/sof/intel/hda-stream.c [hunk 5]
```diff
@@ -85,6 +85,60 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
+	struct sof_intel_dsp_bdl *bdl;
+	int offset, size;
+
+	/* program the initial BDL entries */
+	bdl = (struct sof_intel_dsp_bdl *)stream->bdl.area;
+	offset = 0;
+	stream->frags = 0;
+
+	if (!sdev->first_boot && sdev->trace_init_offset) {
+		/* set up BDL for [trace_init_offset, buffer_size) */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: Can we make this more generic by passing an init_offset and avoid creating two _bdl functions? it's better to deal with the offset and first_boot at a higher level.

File: sound/soc/sof/intel/hda-stream.c [hunk 6]
```diff
@@ -436,7 +491,10 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 
 	hstream->frags = 0;
 
-	ret = hda_dsp_stream_setup_bdl(sdev, dmab, hstream);
+	if (stream_type == SOF_HDA_TRACE_STREAM)
+		ret = hda_dsp_trace_setup_bdl(sdev, dmab, hstream);
+	else
+		ret = hda_dsp_stream_setup_bdl(sdev, dmab, hstream);
 	if (ret < 0) {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: use the same function but pass an explicit offset to setup_bdl. The bdl handling has nothing to do with the trace really, it's generic code

File: sound/soc/sof/intel/hda-trace.c [hunk 7]
```diff
@@ -75,7 +76,6 @@ int hda_dsp_trace_release(struct snd_sof_dev *sdev)
 
 	if (sdev->hda->dtrace_stream) {
 		hstream = &sdev->hda->dtrace_stream->hstream;
-		hstream->opened = false;
 		hda_dsp_stream_put_cstream(sdev,
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-trace.c:79]: why is this removed? Does this belong in this patch?

File: sound/soc/sof/trace.c (hunk 4 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c]: no, trace_init_offset will be refreshed at each resume.
trace_offset is the one come from FW directly.

File: sound/soc/sof/intel/hda-trace.c (hunk 7 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-trace.c:79]: this is actually fixing another issue I met. Let me split it.

File: sound/soc/sof/intel/hda-stream.c (hunk 6 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: hmm, the difference is this:
for trace, we don't have period concept, so BDL entries is not divided to periods, we tried to use only one BDL entry there.
for normal audio stream, we don't have offset!=0 support yet, so for the bdl entry calculation is different, that's why I am adding a new function for trace specific.

but basically I agree with you, let me work on merging them in next version today, thanks.

File: sound/soc/sof/intel/hda-stream.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: As I mentioned in next reply, this first_boot and init_offset logic is not required/implemented/verified on audio streams yet, to reduce regression possibility, I created this separated function for trace.

File: sound/soc/sof/intel/hda-stream.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: OK, let me improve this.

File: sound/soc/sof/trace.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: @keyonjie yes, this makes sense. Sorry about the confusion

**Comment by keyonjie** [comment]
@plbossart @lgirdwood comments addressed and updated, please help review.

File: sound/soc/sof/intel/hda.h [hunk 8]
```diff
@@ -343,6 +343,12 @@ struct sof_intel_dsp_desc {
 	struct snd_sof_dsp_ops *ops;
 };
 
+enum sof_hda_tream_type {
+	SOF_HDA_AUDIO_STREAM = 0,	/* normal playback/capture stream */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.h:347]: "audio" is confusing, we could use this for compressed and probes. Maybe "normal" or just no qualifier.

File: sound/soc/sof/intel/hda-stream.c [hunk 9]
```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
 			     struct snd_dma_buffer *dmab,
-			     struct hdac_stream *stream)
+			     struct hdac_stream *stream, u32 init_offset)
 {
 	struct sof_intel_dsp_bdl *bdl;
-	int i, offset, period_bytes, periods;
-	int remain, ioc;
+	unsigned int i, period_bytes, entries;
+	u32 bytes, offset, size = 0;
+	bool ioc, set_ioc, is_wrap = false;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c:110]: put on separate line if you initialize a variable.

File: sound/soc/sof/intel/hda-stream.c [hunk 10]
```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
+		entries++;
 
-	remain = stream->bufsize % period_bytes;
-	if (remain)
-		periods++;
+	/* count the last index fragment entry */
+	if (init_offset % period_bytes)
+		entries++;
+
+	dev_dbg(sdev->dev, "entries:%d\n", entries);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c:127]: not quite sold on renaming periods to entries? What was the idea here?

File: sound/soc/sof/intel/hda-stream.c [hunk 11]
```diff
@@ -122,19 +137,28 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
...
+					stream, &bdl, offset,
+					bytes, set_ioc && ioc);
+
+		if (offset == stream->bufsize) {
+			offset = 0;
+			is_wrap = true;
+		}
+
+		size += bytes;
 	}
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c:159]: I am unable to determine if the code is correct or not, it's significantly changed from the initial code and I am hesitant to sign-off here. Maybe it's correct, but it's not self-explanatory nor easy to debug/maintain. It'd be easier to deal with logical steps, such as from init_offset to first period start, then all regular periods, then end of buffer alignment, then wrap until last period and again remaining bytes?

File: sound/soc/sof/trace.c [hunk 12]
```diff
@@ -58,11 +63,15 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
...
+		      buffer_size;
+
 	/* return bytes available for copy */
-	if (sdev->host_offset < pos)
+	if (host_offset < pos)
 		return buffer_size - pos;
 	else
-		return sdev->host_offset - pos;
+		return host_offset - pos;
 }
```
Comment (plbossart) [inline sound/soc/sof/trace.c:75]: I don't like copy/paste code with different comments, can this be a helper function?

File: sound/soc/sof/trace.c [hunk 13]
```diff
@@ -154,15 +163,24 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 	if (sdev->dtrace_is_enabled)
 		return -EINVAL;
 
+	/* update init_offset, and reinitial offset to 0 */
```
Comment (plbossart) [inline sound/soc/sof/trace.c:166]: reset offset?

File: sound/soc/sof/trace.c [hunk 14]
```diff
@@ -30,18 +30,23 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 				   loff_t pos, size_t buffer_size)
 {
 	wait_queue_entry_t wait;
+	u32 host_offset;
```
Comment (plbossart) [inline sound/soc/sof/trace.c:33]: this is confusing. You are using 'host_offset but this field was removed from the structures so now it appears as a different concept. did you mean trace_offset?

File: sound/soc/sof/trace.c [hunk 15]
```diff
@@ -154,15 +163,24 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
...
 	params.buffer.size = sdev->dmatb.bytes;
-	params.buffer.offset = 0;
 	params.buffer.pages = sdev->dma_trace_pages;
 
-	sdev->host_offset = 0;
+	/*
+	 * send offset to FW to require copying new logs
+	 * start from new offset if possible.
+	 */
+	params.buffer.offset = sdev->trace_init_offset;
```
Comment (plbossart) [inline sound/soc/sof/trace.c:183]: I am now officially confused. I understand trace_init_offset and the offset to be used on resume. Now you are using it as a dynamic value that's passed to firmware, not sure I get what you are trying to do.

File: sound/soc/sof/trace.c [hunk 16]
```diff
@@ -236,6 +256,8 @@ int snd_sof_init_trace(struct snd_sof_dev *sdev)
 
 	init_waitqueue_head(&sdev->trace_sleep);
 
+	init_waitqueue_head(&sdev->trace_sleep);
```
Comment (plbossart) [inline sound/soc/sof/trace.c:259]: why are we initializing the waitqueue twice?

File: sound/soc/sof/trace.c [hunk 17]
```diff
@@ -252,8 +274,9 @@ EXPORT_SYMBOL(snd_sof_init_trace);
 int snd_sof_trace_update_pos(struct snd_sof_dev *sdev,
 			     struct sof_ipc_dma_trace_posn *posn)
 {
-	if (sdev->dtrace_is_enabled && sdev->host_offset != posn->host_offset) {
-		sdev->host_offset = posn->host_offset;
+	if (sdev->dtrace_is_enabled &&
+	    sdev->trace_offset != posn->host_offset) {
+		sdev->trace_offset = posn->host_offset;
 		wake_up(&sdev->trace_sleep);
```
Comment (plbossart) [inline sound/soc/sof/trace.c:280]: The code is becoming unreadable. Please keep the initial definition of host_offset and init_host_offset, the introduction of 'trace' makes things difficult to follow.

File: sound/soc/sof/intel/hda-stream.c [hunk 18]
```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
-	dev_dbg(sdev->dev, "periods:%d\n", periods);
+	/* count the last offset fragment entry */
+	if (stream->bufsize % period_bytes)
+		entries++;
 
-	remain = stream->bufsize % period_bytes;
-	if (remain)
-		periods++;
+	/* count the last index fragment entry */
+	if (init_offset % period_bytes)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-stream.c:123]: @keyonjie if init_offset is == bufsize, would the number of entries be 2? shouldnt it be just 1?

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h:347]: This was "normal" in initial version, I can change it back if we all aligned, this is not so important to me.

File: sound/soc/sof/intel/hda-stream.c (hunk 9 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:110]: @plbossart I saw you mentioned this somewhere else also, can you share where do you get this recommendation, asking this as I am seeing actually many exceptions in the existed Kernel codes.

File: sound/soc/sof/intel/hda-stream.c (hunk 10 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:127]: I change 'periods' to 'entries' as in Code Loader and Trace cases, we don't have period concept there actually, and this is actually the entries number what we are planning to allocate for BDL entries, it is different with period count, actually.

File: sound/soc/sof/intel/hda-stream.c (hunk 18 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:123]: @ranj063 that won't happen, please refer to:
```
	sdev->trace_init_offset = (sdev->trace_init_offset +
				   sdev->trace_offset) %
				   sdev->dmatb.bytes;
```
init_offset is in range of [0, bufsize).

File: sound/soc/sof/intel/hda-stream.c [hunk 19]
```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
...
+ *
+ * The typical BDL entries after set up should looks like this:
+ *
+ * bdl[0] from the init_offset,  n entries before wrap,
+ * the e-1 full entries, and the last e entry.
+ *
+ *			     init_offset
+ *				 |
+ * |------|...........|---------S-------|------|........|--------|-------|
+ *  bdl[n] bdl[n+j]    bdl[e-1] | bdl[0] bdl[1] bdl[m]   bdl[n-2] bdl[n-1]
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-stream.c:100]: @keyonjie this is so difficult to follow. If I understand it correctly there are only 2 cases:
if init_offset ==0,there are bufsize/periodbytes (+1 in case it is not evenly divisible) periods
else there are 2 periods (or entries as you call it) right?

I am confused with n,m,j and e in the above diagram.

File: sound/soc/sof/intel/hda-stream.c (hunk 18 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:123]: got it!

File: sound/soc/sof/intel/hda-stream.c (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:127]: @keyonjie does "num_buffers" make more sense. It is a list of buffer descriptors after all.

File: sound/soc/sof/intel/hda-stream.c [hunk 20]
```diff
@@ -122,19 +137,28 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
...
-						remain, 0);
+	for (i = 0; i < entries; i++) {
+		bytes = period_bytes - offset % period_bytes;
+		if (is_wrap)
+			bytes = min(bytes, init_offset - offset);
 		else
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						period_bytes, ioc);
+			bytes = min(bytes, stream->bufsize - offset);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-stream.c:145]: @keyonjie I feel like the logic might be wrong here. For the normal stream case where offset is 0, we will end up with "entries" number of buffers of size period_bytes. But the last fragment might not be a full period.

Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:145]: @keyonjie sorry this part is fine.

File: sound/soc/sof/intel/hda-stream.c [hunk 21]
```diff
@@ -122,19 +137,28 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
...
-	for (i = 0; i < periods; i++) {
-		if (i == (periods - 1) && remain)
-			/* set the last small entry */
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						remain, 0);
+	for (i = 0; i < entries; i++) {
+		bytes = period_bytes - offset % period_bytes;
+		if (is_wrap)
+			bytes = min(bytes, init_offset - offset);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-stream.c]: @keyonjie in this case, offset will be 0, so just enough to say:
bytes = min(bytes, init_offset)?

File: sound/soc/sof/intel/hda-stream.c (hunk 11 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:159]: @plbossart I've verified that the logic here is correct. quite impressive @keyonjie 

But might be easier to follow if you break it up like pierre suggests.

File: sound/soc/sof/trace.c (hunk 16 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:259]: sorry, this is refining error with 'patch am', will fix it, thanks.

File: sound/soc/sof/trace.c (hunk 17 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:280]: I agree with Liam that 'trace' is better than 'host' here, as it is member of sdev, but only used for trace(not for other streams).

File: sound/soc/sof/trace.c (hunk 13 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:166]: it's not reset to the same value at each resume, so I am using update...

File: sound/soc/sof/trace.c (hunk 14 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:33]: host_offset is a local variable here to temporary store the latest buffer offset, if it is confusing to you, how about changing it to buffer_offset?

File: sound/soc/sof/intel/hda-stream.c (hunk 19 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:100]: @ranj063 the 2nd case what you mentioned can be more complicated, there can be > 2 periods there.
What the comment I wrote here is aim to abstract all possible cases to this common one. 

You can ignore n, m, j, e here, though they are useful when writing code in logic that Pierre mentioned:

1. from init_offset to first period start: bdl[0]
2. then all regular periods, bdl[m]
3. then end of buffer alignment, bdl[n-1]  ---- We have n items before buffer end;
4. then wrap until last period, bdl[e-2]
5. and again remaining bytes, bdl[e-1]. ---- We have e items in total.

File: sound/soc/sof/intel/hda-stream.c (hunk 11 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:159]: hmm, actually I spent hours to abstract and make it as simple as what it is now, otherwise, it will be inevitable lengthy and low efficiency with those 5 steps Pierre mentioned.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:183]: This is used to cover that case on SKL- platforms where we don't have host DMA, as I mentioned in the PR commit message:

Todo: for SKL- platforms where not using BDL entries, FW to support
non-0 offset need to be implemented for those platforms, please refer
to this:
thesofproject/sof#544

File: sound/soc/sof/trace.c (hunk 12 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:75]: hmm, make sense, let me refine it.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:183]: Then let's be clear: **I will not merge a hardware-agnostic kernel patch that will only work with SKL+ firmware**. Our team has the charter to deal with legacy platforms so the firmware changes need to happen first.

File: sound/soc/sof/intel/hda-stream.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:110]: it's not because others did things poorly in the past that they should be followed. The recommended practice is a single variable per line and x-mas tree shape.

File: sound/soc/sof/intel/hda-stream.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:127]: bdl_entries then...

File: sound/soc/sof/intel/hda-stream.c (hunk 11 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:159]: You may have spent hours but I have no way of knowing if it works or not. 
How many times have we faced issues with corner cases in this sort of code?
Please.

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:347]: yes, normal. 
And fix typo above, it's sof_hda_stream_type

File: sound/soc/sof/intel/hda.h [hunk 22]
```diff
@@ -442,14 +448,15 @@ void hda_dsp_stream_free(struct snd_sof_dev *sdev);
 int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 			     struct hdac_ext_stream *stream,
 			     struct snd_dma_buffer *dmab,
-			     struct snd_pcm_hw_params *params);
+			     struct snd_pcm_hw_params *params,
+			     enum sof_hda_tream_type stream_type);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.h:452]: and typo here as well

File: sound/soc/sof/trace.c (hunk 13 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:166]: my point was that 'reinitial' is not an English word.

File: sound/soc/sof/trace.c (hunk 17 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:280]: then change the dma_trace_posn structure to use trace_offset... It makes no sense to use different fields in different structures when it's the same concept.

File: sound/soc/sof/trace.c (hunk 14 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:33]: it helps if the local variable has the same name as the field in the structure you compare against.e.g.
````c
trace_offset = (sdev->trace_init_offset + sdev->trace_offset) %
		      buffer_size;
````

otherwise, you might as well call it foobar

File: sound/soc/sof/intel/hda-stream.c (hunk 19 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:100]: @keyonjie Yes, I understand the logic and ther is not doubt it is correct.

If I may make a suggestion: can we treat the case of a non-zero init_offset as 2 separate buffers,
then we end up with just as many bld entries except that the partial segments will be only at the end of the 2 buffer areas. For example: 
Case 1: init_offset = 0
num of period_bytes size bdl entries = 5
num of partial bdl entries = 1

init_offset = 0
|
|------|------|------|------|------|---|

Case 2: init_offset = 16
num of period_bytes size bdl entries = 4
num of partial bdl entries = 2

init_offset = 16
                      
|------|------|----|------|------|-----|

So the partial fragments are at the end of the 2 areas ( 0 - init_offset -1, init_offset - bufsize).

But with the current case there will be 3 partial entries. 

Anyway, this is only to improve maintainability. I am fine if the code is merged as is too.

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h:347]: OK, will change.

File: sound/soc/sof/intel/hda-stream.c (hunk 9 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:110]: OK, let me change to follow this, good to know about this, thanks.

File: sound/soc/sof/intel/hda-stream.c (hunk 10 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:127]: OKay, will change.

File: sound/soc/sof/intel/hda-stream.c (hunk 11 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:159]: So, shall I try add more comments with logic analyzing?

File: sound/soc/sof/trace.c (hunk 13 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:166]: Okay, changing it to 're-initialize'.

File: sound/soc/sof/trace.c (hunk 14 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:33]: OK, let me change it to trace_offset.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:183]: hmm, this was the biggest concern from me when writing this patch, let me raise #544 and try if we can speed up the implementation in FW side.

File: sound/soc/sof/intel/hda.h (hunk 22 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h:452]: thanks, will change.

File: sound/soc/sof/trace.c (hunk 17 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:280]: the dma_trace_posn struct is replied from FW, looks host_offset better there. the sdev is only used in driver, so they have actually different standpoint, can we just keep this?

File: sound/soc/sof/intel/hda-stream.c (hunk 19 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:100]: > @keyonjie Yes, I understand the logic and ther is not doubt it is correct.
> 
> If I may make a suggestion: can we treat the case of a non-zero init_offset as 2 separate buffers,
> then we end up with just as many bld entries except that the partial segments will be only at the end of the 2 buffer areas. For example:
> Case 1: init_offset = 0
> num of period_bytes size bdl entries = 5
> num of partial bdl entries = 1
> 
> init_offset = 0
> |
> |------|------|------|------|------|---|
> 
> Case 2: init_offset = 16
> num of period_bytes size bdl entries = 4
> num of partial bdl entries = 2
> 
> init_offset = 16
> 
> |------|------|----|------|------|-----|
> 
> So the partial fragments are at the end of the 2 areas ( 0 - init_offset -1, init_offset - bufsize).
> 
> But with the current case there will be 3 partial entries.

Sorry I don't catch you here, there will be only 2 partial entries in your sample.

> 
> Anyway, this is only to improve maintainability. I am fine if the code is merged as is too.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:183]: Something is wrong with your references, torvalds#544 does not refer to any firmware issues

File: sound/soc/sof/trace.c (hunk 17 above)
Comment (plbossart) [reply sound/soc/sof/trace.c:280]: I don't understand your reply. Why would the same concept be named differently if this is firmware or driver?

File: sound/soc/sof/intel/hda-stream.c (hunk 11 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:159]: The ask is to find a way to make things self-explanatory with all corner cases handled. You can do this with simple code or comments, your choice.

File: sound/soc/sof/intel/hda-stream.c (hunk 19 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-stream.c:100]: @keyonjie what I meant was just break up the stream->bufsize into 2 and then assign bdl entries individually for [init_offset, buf_size) and then [0, init_offset -1). This way the logic will be a lot easier to follow. Let me take a swipe at it today and send you a patch

Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:100]: @ranj063 the init_offset may be used for normal audio stream also, so we still need assign bdl entries for periods when init_offset!=0.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c:183]: sorry, it should be thesofproject/sof#544.
by the way, @plbossart , just synced with @ranj063 , we actually haven't enable PM on SKL- platforms yet, so init_offset there are always 0, we this PR won't impact the current behaviors on those platforms.

**Comment by keyonjie** [comment]
Let me close this long conversation PR, split and create new PRs.

**Comment by keyonjie** [comment]
update the status of logger issues:
after #287 merged,  issue #271 and thesofproject/sof#526 are fixed, @ranj063 please backport it to cros branch.
after manually applied #287 and #237, https://github.com/thesofproject/sof/issues/401 is already fixed on my observation.

The left logger issue is: the trace log before suspend will be cleared after resume, this is known limitation in our current design, the issue is not so urgent and no complaint about it yet, the bdl refining part in this PR(#248) can fix that, I will refine and send new PR for it.
